File: web/studio/ASC.Web.Studio/UserControls/Common/TabsNavigator/js/tabsnavigator.js
/* 
 * 
 * (c) Copyright Ascensio System Limited 2010-2014
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * http://www.gnu.org/licenses/agpl.html 
 * 
 */

if (typeof(ASC) == "undefined") {
    ASC = {};
}
if (typeof(ASC.Controls) == "undefined") {
    ASC.Controls = {};
}

ASC.Controls.TabsNavigator = new function () {
    (function () {
        ASC.Controls.AnchorController.bind('onupdate', function () {
            var anchor = ASC.Controls.AnchorController.getAnchor();
            try {
                var $tabObj = jq(".tabsNavigationLinkBox>.tabsNavigationLink_" + anchor + ":first");
                if ($tabObj) {
                    ASC.Controls.TabsNavigator.toggleTabs(jq($tabObj).attr("id"), "");
                }
            } catch (e) {
            }
        },
            { 'once': true }
        );
    })();

    var toggleCurrentTab = function (tab, tabID) {
        var hideFlag = true;
        var currentTabID = tab.attr("id");

        if (tab.hasClass("selectedTab")) {
            tab.removeClass("selectedTab");
        }
        if (currentTabID == tabID) {
            tab.addClass("selectedTab");
            hideFlag = false;
        }

        if (currentTabID == undefined || currentTabID == '' || currentTabID == null) {
            currentTabID = '';
        }
        var currentDivID = currentTabID.replace(/_tab$/gi, '');

        if (currentDivID != '' && jq("#" + currentDivID).length == 1) {
            if (hideFlag) {
                jq("#" + currentDivID + ":not(.display-none)").addClass("display-none");
            } else {
                jq("#" + currentDivID + ".display-none").removeClass("display-none");
            }
        }
    };

    this.toggleTabs = function (tabID, anchor) {
        var tab = jq("#" + tabID);
        tab.parent().children('a').each(
            function () {
                var child = jq(this);
                toggleCurrentTab(child, tabID);
            }
        );
        if (typeof(anchor) == "string" && anchor != "") {
            ASC.Controls.AnchorController.move(anchor);
        }
    };
};

File: web/studio/ASC.Web.Studio/UserControls/Common/TabsNavigator/TabsNavigator.cs
/* 
 * 
 * (c) Copyright Ascensio System Limited 2010-2014
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * http://www.gnu.org/licenses/agpl.html 
 * 
 */

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Text;
using System.Web;
using System.Web.UI;
using System.Linq;

namespace ASC.Web.Studio.UserControls.Common.TabsNavigator
{
    [
        ToolboxData("<{0}:TabsNavigator runat=\"server\"/>"),
        ParseChildren(ChildrenAsProperties = true), PersistChildren(true)
    ]
    public class TabsNavigator : Control
    {
        #region Properties

        public string BlockID { get; set; }


        private List<TabsNavigatorItem> _tabItems;

        [Description("List of tabs."),
         Category("Data"), PersistenceMode(PersistenceMode.InnerProperty),
         DesignerSerializationVisibility(DesignerSerializationVisibility.Content), NotifyParentProperty(true)
        ]
        public List<TabsNavigatorItem> TabItems
        {
            get { return _tabItems ?? (_tabItems = new List<TabsNavigatorItem>()); }
            set { _tabItems = value; }
        }

        protected bool HasTabItems
        {
            get { return TabItems.Count > 0; }
        }

        #endregion

        protected override void OnLoad(EventArgs e)
        {
            foreach (var tab in TabItems.Where(tab => tab.Visible))
            {
                Controls.Add(tab);
            }
            base.OnInit(e);
            InitScripts();
        }

        protected void InitScripts()
        {
            Page.RegisterBodyScripts(ResolveUrl("~/usercontrols/common/tabsnavigator/js/tabsnavigator.js"));
        }

        protected override void Render(HtmlTextWriter writer)
        {
            var sb = new StringBuilder();

            sb.Append("<div class=\"clearFix\">");
            sb.Append("  <div id=\"" + (String.IsNullOrEmpty(BlockID) ? ClientID : BlockID) + "\" class=\"tabsNavigationLinkBox\">");

            if (HasTabItems)
            {
                var visibleTabItems = TabItems.Where(tab => tab.Visible).ToList();
                var visibleTabsCount = visibleTabItems.Count;

                for (var i = 0; i < visibleTabsCount; i++)
                {
                    sb.Append(visibleTabItems[i].GetTabLink(i == visibleTabsCount - 1));
                }
            }

            sb.Append("  </div>");
            sb.Append("</div>");

            writer.Write(sb.ToString());

            foreach (var tab in TabItems.Where(tab => tab.Visible && string.IsNullOrEmpty(tab.TabHref) && !tab.SkipRender))
            {
                tab.RenderTabContent(writer);
            }
        }
    }
}